Lib/test/test_urlparse.py

```diff
@@ -678,6 +734,38 @@ def test_urljoins_relative_base(self):
             self.checkJoin('http://a', scheme + '/w', 'http://a/w')
             self.checkJoin('http://a', scheme + '///w', 'http://a/w')
             self.checkJoin('http://a', scheme + 'w', 'http://a/w')
+            self.checkJoin('http://a', scheme + '../w', 'http://a/w')
+            self.checkJoin('http://a', scheme + './w', 'http://a/w')
+            self.checkJoin('http://a', scheme + '..//w', 'http://a/w')
+            self.checkJoin('http://a', scheme + './/w', 'http://a/w')
+            self.checkJoin('http://a', scheme + '..///w', 'http://a//w')
+            self.checkJoin('http://a', scheme + './//w', 'http://a//w')
+            self.checkJoin('http://a', scheme + '..', 'http://a')
+            self.checkJoin('http://a', scheme + '.', 'http://a')
+            self.checkJoin('http://a', scheme + '../', 'http://a')
+            self.checkJoin('http://a', scheme + './', 'http://a')
+            self.checkJoin('http://a', scheme + '..//', 'http://a/')
+            self.checkJoin('http://a', scheme + './/', 'http://a/')
+            self.checkJoin('http://a', scheme + '..///', 'http://a//')
+            self.checkJoin('http://a', scheme + './//', 'http://a//')
+
+        self.checkJoin('b/c', '', 'b/c')
+        self.checkJoin('b/c', '//', 'b/c')
+        self.checkJoin('b/c', '//v', '//v')
+        self.checkJoin('b/c', '//v/w', '//v/w')
+        self.checkJoin('b/c', '/w', '/w')
+        self.checkJoin('b/c', '///w', '/w')
+        self.checkJoin('b/c', 'w', 'b/w')
+        self.checkJoin('b/c', '../w', 'w')
+        self.checkJoin('b/c', '../../w', 'w')
+        self.checkJoin('b/c', '../../../w', 'w')
+        self.checkJoin('b/c', 'w/.', 'b/w/')
+        self.checkJoin('b/c', '../w/.', 'w/')
+        self.checkJoin('b/c', '../../w/.', 'w/')
+        self.checkJoin('b/c', '../../../w/.', 'w/')
+        self.checkJoin('b/c', '..', '')
+        self.checkJoin('b/c', '../..', '')
+        self.checkJoin('b/c', '../../..', '')
```

Comment:
Although these fall outside the direct scope of the pseudocode defined in RFC 3986 because `b/c` is not an absolute base URI, they violate the obvious expectation that `urljoin` should be associative. See

- #96015

Given non–RFC 3986 input where the base URI is path-relative (undefined scheme, undefined authority, and path not beginning with `/`), we should preserve extra initial `..` components in the output:

```suggestion
        self.checkJoin('b/c', '../../w', '../w')
        self.checkJoin('b/c', '../../../w', '../../w')
        self.checkJoin('b/c', 'w/.', 'b/w/')
        self.checkJoin('b/c', '../w/.', 'w/')
        self.checkJoin('b/c', '../../w/.', '../w/')
        self.checkJoin('b/c', '../../../w/.', '../../w/')
        self.checkJoin('b/c', '..', '')
        self.checkJoin('b/c', '../..', '..')
        self.checkJoin('b/c', '../../..', '../..')
```